### DISCUSSION TITLE
Practical differences between keyframe cut normal cut? Direct ffmpeg backend?

### DISCUSSION THREAD

**Discussion by rieje** [mifi/lossless-cut#1081] [Q&A]
On the [README](https://github.com/mifi/lossless-cut#known-issues--limitations), it implies a difference between keyframe cut and normal cut and also an ffmpeg documentation that kind of goes over the implementation of the two (i.e. input seeking vs output seeking).

But between the link the only thing mentioned is, from the ffmpeg documentation, 

> As of FFmpeg 2.1, the main advantage is that when applying filters to the output stream, the timestamps aren't reset prior to filtering (i.e. when ​burning subtitles into a video, you don't need to modify the subtitle timestamps), but the drawback is that it will take a lot of time until it finally reaches that time point. The bigger the seeking time is, the longer you will have to wait. 

I simply use lossless-cut to cut/join videos losslessly, without modifying metadata, subtitle timestamps, etc. When should one be used over the other and how would this affect the results of the processed video? My intuition is that keyframe cut will always produce videos with no problems except that timestamps aren't exact because it cuts at keyframes, while normal cut will respect timestamps but the resulting video may have problems. If this is correct, what are examples of problems a normal cut may produce?

Also, not related to lossless-cut, but I was looking to automate the process of cutting/joining for a more efficient workflow using mpv. lossless-cut will be useful because of its developing smart cut feature which as I understand is as lossless as you can get while being able to cut to the exact frame (it only re-encodes the surrounding frames), but I was wondering if keyframe cut vs. normal cut in lossless-cut is as simple as:

    # -ss before -i for input seeking, i.e. keyframe cut, guaranteed to produce video with no problems but timestamps likely not exact 
    ffmpeg  -ss 0:0:4 -i originalvideo.mp4 -t 0:1:10 -vcodec copy -acodec copy outputvideo.mp4
    # vs. 
    # -ss after -i for output seeking, i.e. normal cut, guaranteed to produce video with exact timestamps but video likely to have "problems" 
   ffmpeg -i originalvideo.mp4 -ss 0:0:4 -t 0:1:10 -vcodec copy -acodec copy outputvideo.mp4

?

If so, then I would rather use a wrapper script to save timestamps of videos within mpv (easily doable) and process that through ffmpeg wrapper script for an efficient workflow, and use lossless-cut for smart cuts if the video is supported. Otherwise, if keyframe cut or normal cut in lossless-cut does more than the commands above, then I would pass the saved timestamps as a .csv and import it to losless-cut, which would be less efficient.

Would love to know if there are any other relevant ffmpeg settings one should know for my purposes. Much appreciated for the great software and love that smart cut is being worked on as it's a feature that's almost exclusively found only in proprietary applications.

**Comment by mifi** [comment]
> Also, not related to lossless-cut, but I was looking to automate the process of cutting/joining for a more efficient workflow using mpv. lossless-cut will be useful because of its developing smart cut feature which as I understand is as lossless as you can get while being able to cut to the exact frame (it only re-encodes the surrounding frames), but I was wondering if keyframe cut vs. normal cut in lossless-cut is as simple as:
> 
> ```
> # -ss before -i for input seeking, i.e. keyframe cut, guaranteed to produce video with no problems but timestamps likely not exact 
> ffmpeg  -ss 0:0:4 -i originalvideo.mp4 -t 0:1:10 -vcodec copy -acodec copy outputvideo.mp4
> # vs. 
> # -ss after -i for output seeking, i.e. normal cut, guaranteed to produce video with exact timestamps but video likely to have "problems" 
> ```
> 
> ffmpeg -i originalvideo.mp4 -ss 0:0:4 -t 0:1:10 -vcodec copy -acodec copy outputvideo.mp4
> 
> ?

Yes, the only difference between keyframe and normal cut is whether the `-ss x` argument is placed before or after `-i`.

As for the consequences of using one or the other, there has been many discussions including #330 #13 but no clear conclusion can be made. It depends on the particular file in question. I have not found any 100% deterministic behavior, other than "keyframe cut" **usually** yields better results than "normal cut" (but not in all cases)
